src/sentry/snuba/metrics/extraction.py

```diff
@@ -49,6 +52,66 @@
 
 logger = logging.getLogger(__name__)
 
+
+# This helps us control the different spec versions
+# in order to migrate customers from invalid specs
+class SpecVersion(NamedTuple):
+    version: int
+    flags: set[str] = set()
+
+    def matches_flags_set(self, flags: set[str]) -> bool:
+        return flags == self.flags
+
+
+class OnDemandMetricSpecVersioning:
+    # The flags from the spec versions alter the generated metric in OnDemandMetricSpec
+    # Once we're ready to abandon a version we will coalesce the list into a single element
+    # When there's a single version we should not have any flags and we should change
+    # OnDemandMetricSpec accordingly
+    spec_versions = [
+        SpecVersion(0),
+        SpecVersion(1, {"use_updated_env_logic"}),
+    ]
+
+    feature_to_flags_map = {
+        "organizations:on-demand-query-with-new-env-logic": {"use_updated_env_logic"},
+    }
+
+    @classmethod
+    def get_spec_versions(cls: Any) -> Sequence[SpecVersion]:
+        """Get all spec versions."""
+        return cls.spec_versions
+
+    @classmethod
+    def get_default_spec_version(cls: Any) -> SpecVersion:
+        # I believe this should be the original behaviour rather than the new one,
+        # however, the current code defaults to the new `use_updated_env_logic: bool = True``
+        return cls.spec_versions[-1]
+
+    @classmethod
+    def get_query_spec_version(cls: Any, feature_flag: str, org_id: int) -> SpecVersion:
+        """Return spec version based on feature flag enabled for an organization."""
+        org = Organization.objects.get_from_cache(id=org_id)
+        flags_set = (
+            cls.feature_to_flags_map[feature_flag] if features.has(feature_flag, org) else []
+        )
+
+        return cls._find_spec_version(flags_set)
+
+    @classmethod
+    def get_query_spec_version_flags_set(cls: Any, flags_set: set[str]) -> SpecVersion:
+        """Return spec version matching flags set."""
+        return cls._find_spec_version(flags_set)
+
+    @classmethod
+    def _find_spec_version(cls: Any, flags_set: set[str]) -> SpecVersion:
+        return [
+            spec_version
+            for spec_version in cls.spec_versions
+            if spec_version.matches_flags_set(flags_set)
+        ][0]
```

Comment:
It's a pretty risky piece of code since it will throw an error if no specs are found. Maybe more explicit error handling should be done (in case we want to proceed with this more complex impl).